docs/userguide/mpi_apps.rst

```diff
@@ -40,116 +73,275 @@ to the desired number of blocks.
+                        account=user_opts["account"],
+                        queue=user_opts["queue"],
+                        # PBS directives (header lines): for array jobs pass '-J' option
+                        scheduler_options=user_opts["scheduler_options"],
+                        # Command to be run before starting a worker, such as:
+                        worker_init=user_opts["worker_init"],
+                        # number of compute nodes allocated for each block
+                        nodes_per_block=user_opts["nodes_per_block"],
+                        init_blocks=1,
+                        min_blocks=0,
+                        max_blocks=1, # Can increase more to have more parallel jobs
+                        walltime=user_opts["walltime"]
+                    ),
                 ),
-            ),
-        ]
-    )
+            ],
 
 
 Writing MPI-Compatible Apps
 ++++++++++++++++++++++++++++
 
-The `App <apps.html>`_ can be either a Python or Bash App which invokes the MPI application.
-
-In the easiest case (i.e., single MPI task per block), write the MPI launcher options in the string returned by
-the bash app or as part of a subprocess call from a Python app.
+In MPI mode, the :class:`~parsl.executors.high_throughput.executor.HighThroughputExecutor` can execute both Python or Bash Apps which invokes the MPI application.
+However, it is important to not that Python Apps that directly use ``mpi4py`` is not supported.
 
+For multi-node MPI applications, especially when running multiple applications within a single batch job,
+it is important to specify the resource requirements for the app so that the Parsl worker can provision
+the appropriate resources before the application starts. For eg, your Parsl script might contain a molecular
+dynamics application that requires 8 ranks over 1 node for certain inputs and 32 ranks over 4 nodes for some
+depending on the size of the molecules being simulated. By specifying resources via ``resource_specification``,
+parsl workers will provision the requested resources and then compose MPI launch command prefixes
+(Eg: ``mpiexec -n <ranks> -ppn <ranks_per_node> -hosts <node1..nodeN>``). These launch command prefixes are
+shared with the app via environment variables.
 
 .. code-block:: python
 
     @bash_app
-    def echo_hello(n: int, stderr='std.err', stdout='std.out'):
-        return f'mpiexec -n {n} --ppn 1 hostname'
+    def echo_hello(n: int, stderr='std.err', stdout='std.out', parsl_resource_specification: Dict):
+        return f'$PARSL_MPI_PREFIX hostname'
+
+    # Alternatively, you could also use the resource_specification to compose a launch
+    # command using env vars set by Parsl from the resource_specification like this:
+    @bash_app
+    def echo_something(n: int, stderr='std.err', stdout='std.out', parsl_resource_specification: Dict):
+        total_ranks = os.environ("")
+        return f'aprun -N $PARSL_RANKS_PER_NODE -n hostname'
```

Comment:
Where does the `hostname` come from in this example? What are the environment variables which Parsl sets?

Reply:
Clarified in docs. The idea here was for each rank to print the output from `/bin/hostname`.

Reply:
Is the syntax wrong then? Should there not be an integer after `-n`.

Reply:
Fixed. The app looks like this now:

```python
@bash_app
def echo_hostname(n: int, stderr='std.err', stdout='std.out', parsl_resource_specification: Dict):
    total_ranks = os.environ("")
    return f'aprun -N $PARSL_RANKS_PER_NODE -n {total_ranks} /bin/hostname'
```

Reply:
Fixed.